worker/worker.go

```diff
@@ -0,0 +1,239 @@
+// called. If the caller fails to call either of these methods and calls Stop()
+// in the Worker, then that method will block forever.
+func (w *Worker) Tasks() <-chan *Task {
+	return w.tasks
+}
+
+// Lifecycle returns the Lifecycle used to Complete or Abandon an item from the
+// Queue. It makes use of the Mutex guarding the Lifecycle and returns a default
+// value when no Lifecycle is set.
+func (w *Worker) Lifecycle() Lifecycle {
+	w.lmu.RLock()
+	defer w.lmu.RUnlock()
+
+	if w.lifecycle == nil {
+		panic("TODO(worker): w.lifecycle is nil")
+	}
+
+	return w.lifecycle
+}
+
+// SetLifecycle blocks until a write lock is available on the guarding mutex,
+// and then updates it. The next item processed by the queue will be handled
+// using the new Lifecycle passed in as `l`.
+func (w *Worker) SetLifecycle(l Lifecycle) {
+	w.lmu.Lock()
+	defer w.lmu.Unlock()
+
+	w.lifecycle = l
+}
+
+// Close stops polling the queue immediately and waits for all tasks to complete
+// before stopping the heartbeat.
+func (w *Worker) Close() {
+	w.queue.Close()
+
+	w.closer <- struct{}{}
+	w.wg.Wait()
+
+	w.heart.Close()
+}
+
+// Halt stops the heartbeat and queue polling goroutines immediately and cancels
+// all tasks, marking them as FAILED before returning.
+func (w *Worker) Halt() {
+	w.closer <- struct{}{}
+
+	w.queue.Close()
+	w.heart.Close()
+
+	for i := 0; i < int(w.taskCount); i++ {
+		w.halter <- struct{}{}
+	}
+}
+
+// recv begins processing elements from the incoming queue, and delegating them
+// as Tasks along the wire. If an error occurs at the queue level, it will be
+// propagated up the chain to the Errs() channel on the Worker. Otherwise, if a
+// close or halt operation is given, recv will return immediately.
+//
+// This function runs in its own goroutine.
+func (w *Worker) recv() {
+	for {
+		select {
+		case payload := <-w.queue.In():
+			task := NewTask(payload, w.monitor(payload))
+			task.Begin()
+
+			w.tasks <- task
+		case err := <-w.queue.Errs():
+			w.errs <- err
+		case <-w.closer:
+			return
+		}
+	}
+}
+
+// monitor creates and returns a channel used to monitor when the Task is
+// marked as SUCCESSFUL or FAILED. Upon receiving a status update, it will
+// either call the Complete or Abandon function on the lifecycle. If a call to
+// Halt() is dispatched, the lifecycle will immediately abandon the current
+// task.
+func (w *Worker) monitor(payload []byte) (progress chan<- bool, closer <-chan struct{}) {
+	w.incrementTaskCount(1)
+
+	lifecycle := w.Lifecycle()
+	progress = make(chan bool)
+	closer = make(chan struct{})
+
+	go func(progress <-chan bool) {
+		defer w.decrementTaskCount()
+		defer close(progress)
+
+		var err error
+
+		select {
+		case success := <-progress:
+			if success {
+				err = lifecycle.Complete(payload)
+			} else {
+				err = lifecycle.Abandon(payload)
+			}
+		case <-w.halter:
+			closer <- struct{}{}
+			err = lifecycle.Abandon(payload)
```

Comment:
@connor4312 This is a line that I was having particular difficulty with. I'm not sure whether or not we need to let a Task know that its progress has been halted. The main motivation behind it was if a, say, `ONLINE` event was being processed, but was halted, should the Task stop updating state-ful services? Perhaps a state could even roll back the changes with the addition of a state machine? I'm not sure if this is the sort of behavior that we want, but I figured I'd open the discussion.

Curious as to your thoughts.